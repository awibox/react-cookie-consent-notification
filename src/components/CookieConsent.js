import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { LocalStorage } from 'combo-storage';

const CookieConsentStyle = {
  position: 'fixed',
  display: 'flex',
  flexFlow: 'row',
  alignItems: 'center',
  justifyContent: 'space-between',
  right: 0,
  left: 0,
  boxShadow: '0 0 3px rgba(0, 0, 0, 0.12), 0 0 2px rgba(0, 0, 0, 0.24)',
  zIndex: 900,
};

const CookieConsentButtonStyle = {
  borderRadius: '3px',
  boxShadow: '0 0 3px rgba(0, 0, 0, 0.12), 0 0 2px rgba(0, 0, 0, 0.24)',
  whiteSpace: 'nowrap',
  cursor: 'pointer',
};

class CookieConsent extends Component {
  constructor(props) {
    super(props);
    this.state = {
      close: false,
      hasStorage: !!LocalStorage.get('react-cookie-consent-notification'),
    };
    this.consentAction = this.consentAction.bind(this);
  }

  static propTypes = {
    background: PropTypes.string,
    bottomPosition: PropTypes.bool,
    buttonText: PropTypes.string,
    buttonBackground: PropTypes.string,
    buttonColor: PropTypes.string,
    buttonFontSize: PropTypes.number,
    buttonClassName: PropTypes.string,
    children: PropTypes.any,
    className: PropTypes.string,
    color: PropTypes.string,
    consentFunction: PropTypes.func,
    padding: PropTypes.number,
  };

  static defaultProps = {
    background: '#fff',
    bottomPosition: false,
    buttonText: 'I agree',
    buttonBackground: '#fff',
    buttonColor: '#000',
    buttonFontSize: 16,
    children: 'This website uses cookies to improve service, for analytical and advertising purposes.',
    color: '#000',
    consentFunction: () => {},
    padding: 20,
  };

  consentAction() {
    this.setState({ close: true });
    LocalStorage.set('react-cookie-consent-notification', true);
    if (!this.props.bottomPosition) {
      document.body.style.marginTop = 0;
    }
    this.props.consentFunction(true);
  }

  componentDidMount() {
    if (!this.state.hasStorage && !this.props.bottomPosition) {
      const consent = document.getElementById('cookie-consent').offsetHeight;
      document.body.style.marginTop = `${consent}px`;
    }
    this.props.consentFunction(this.state.hasStorage);
  }

  render() {
    if (this.state.hasStorage) {
      return null;
    }
    const {
      background,
      bottomPosition,
      buttonText,
      buttonBackground,
      buttonColor,
      buttonFontSize,
      buttonClassName,
      children,
      className,
      color,
      padding,
    } = this.props;
    const consentAction = this.state.close;
    const positionObj = bottomPosition ? { bottom: 0 } : { top: 0 };
    const styleForCookieConsent = {
      visibility: consentAction ? 'hidden' : 'visible',
      backgroundColor: background,
      padding: `${padding / 2}px`,
      color,
      ...positionObj,
      ...CookieConsentStyle,
    };
    const styleForCloseButton = {
      backgroundColor: buttonBackground,
      color: buttonColor,
      height: `${buttonFontSize}px`,
      lineHeight: `${buttonFontSize}px`,
      fontSize: `${buttonFontSize}px`,
      margin: `0 ${padding / 2}px`,
      padding: `${padding / 2}px ${padding}px`,
      ...CookieConsentButtonStyle,
    };
    return (
      <div id='cookie-consent' className={className} style={className ? {} : styleForCookieConsent}>
        <div style={{ padding: `${padding / 2}px` }}>{children}</div>
        <div id='cookie-consent-button'
             onClick={this.consentAction}
             className={buttonClassName}
             style={buttonClassName ? {} : styleForCloseButton}>{buttonText}</div>
      </div>
    );
  }
}
export default CookieConsent;
