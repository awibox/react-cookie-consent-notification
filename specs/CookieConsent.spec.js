import React from 'react';
import { shallow } from 'enzyme';
import CookieConsent from '../src/components/CookieConsent';

const testColor = '#fff';
const testBackground = '#000';
const testText = 'This is correct text';
const testCustomContent = 'Custom content';
const testClassName = 'customClass';

describe('CookieConsent', () => {
  it('should render correctly', () => {
    const component = shallow(<CookieConsent/>);
    expect(component).toMatchSnapshot();
  });
  it('should not render', () => {
    const component = shallow(<CookieConsent>{testText}</CookieConsent>);
    component.setState({ hasStorage: true });
    expect(component.get(0)).toEqual(null);
  });
  it('should have background', () => {
    const component = shallow(<CookieConsent background={testBackground}/>);
    expect(component.find('#cookie-consent').get(0).props.style.backgroundColor).toEqual(testBackground);
  });
  it('should have text color', () => {
    const component = shallow(<CookieConsent color={testColor}/>);
    expect(component.find('#cookie-consent').get(0).props.style.color).toEqual(testColor);
  });
  it('should have bottom position', () => {
    const component = shallow(<CookieConsent bottomPosition={true}/>);
    expect(component.find('#cookie-consent').get(0).props.style.bottom).toEqual(0);
  });
  it('should have top position', () => {
    const component = shallow(<CookieConsent/>);
    expect(component.find('#cookie-consent').get(0).props.style.top).toEqual(0);
  });
  it('should close notification', () => {
    const component = shallow(<CookieConsent/>);
    component.find('#cookie-consent-button').simulate('click');
    expect(component.find('#cookie-consent').get(0).props.style.visibility).toEqual('hidden');
  });
  it('should close notification bottom position', () => {
    const component = shallow(<CookieConsent bottomPosition={true}/>);
    component.find('#cookie-consent-button').simulate('click');
    expect(component.find('#cookie-consent').get(0).props.style.visibility).toEqual('hidden');
  });
  it('should have a className', () => {
    const component = shallow(<CookieConsent className={testClassName}>
      {testCustomContent}
    </CookieConsent>);
    expect(component.find(`.${testClassName}`)).toExist();
  });
  it('should have a className', () => {
    const component = shallow(<CookieConsent className={testClassName}>
      {testCustomContent}
    </CookieConsent>);
    expect(component.find(`.${testClassName}`)).toExist();
  });
  it('should have a button className', () => {
    const component = shallow(<CookieConsent buttonClassName={testClassName}>
      {testCustomContent}
    </CookieConsent>);
    expect(component.find('#cookie-consent-button')).toHaveClassName(testClassName);
  });
});
