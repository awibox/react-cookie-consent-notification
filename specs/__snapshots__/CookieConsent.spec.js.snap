// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieConsent should render correctly 1`] = `
<div
  id="cookie-consent"
  style={
    Object {
      "backgroundColor": "#fff",
      "boxShadow": "0 0 3px rgba(0, 0, 0, 0.12), 0 0 2px rgba(0, 0, 0, 0.24)",
      "color": "#000",
      "left": 0,
      "padding": "20px",
      "paddingRight": "48px",
      "position": "fixed",
      "right": 0,
      "top": 0,
      "visibility": "visible",
      "zIndex": 900,
    }
  }
>
  <div
    id="cookie-consent-close"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "fontFamily": "-webkit-pictograph",
        "fontSize": "28px",
        "height": "28px",
        "lineHeight": "28px",
        "marginTop": "-14px",
        "position": "absolute",
        "right": "10px",
        "textAlign": "center",
        "top": "50%",
        "transform": "rotate(45deg)",
        "width": "28px",
      }
    }
  >
    +
  </div>
</div>
`;
